pysmt/rewritings.py

```diff
@@ -0,0 +1,353 @@
+                new_q_vars = set(q_vars)
+                if len(needs_rename) > 0:
+                    sub = {v : self.mgr.FreshSymbol(v.symbol_type())
+                           for v in needs_rename}
+                    sub_matrix = sub_matrix.substitute(sub)
+                    new_q_vars = (q_vars - needs_rename)
+                    new_q_vars |= set(sub[x] for x in needs_rename)
+                sq.append((Q, new_q_vars))
+                reserved |= new_q_vars
+
+            quantifiers += sq
+            matrix.append(sub_matrix)
+
+        if formula.is_and():
+            return (quantifiers, self.mgr.And(matrix))
+        if formula.is_or():
+            return (quantifiers, self.mgr.Or(matrix))
+
+    def walk_not(self, formula, args):
+        quantifiers, matrix = args[0]
+
+        nq = [(self._invert_quantifier(Q), qvars) for Q, qvars in quantifiers]
+        return (nq, self.mgr.Not(matrix))
+
+    def walk_iff(self, formula, args):
+        a, b = formula.args()
+        i1 = self.mgr.Implies(a, b)
+        i2 = self.mgr.Implies(b, a)
+        i1_args = self.walk_implies(i1, args)
+        i2_args = self.walk_implies(i2, list(reversed(args)))
+        return self.walk_conj_disj(self.mgr.And(i1, i2), [i1_args, i2_args])
+
+    def walk_implies(self, formula, args):
+        a, b = formula.args()
+        na = self.mgr.Not(a)
+        na_arg = self.walk_not(na, [args[0]])
+        return self.walk_conj_disj(self.mgr.Or(na, b), [na_arg, args[1]])
+
+    def walk_ite(self, formula, args):
+        if any(a is None for a in args):
+            return None
+        else:
+            i, t, e = formula.args()
+            i_args, t_args, e_args = args
+            ni = self.mgr.Not(i)
+            i1 = self.mgr.Implies(i, t)
+            i2 = self.mgr.Implies(ni, e)
+            ni_args = self.walk_not(ni, [i_args])
+            i1_args = self.walk_implies(i1, [i_args, t_args])
+            i2_args = self.walk_implies(i2, [ni_args, e_args])
+            return self.walk_conj_disj(self.mgr.And(i1, i2), [i1_args, i2_args])
+
+    def walk_theory_relation(self, formula, args):
+        return [], formula
+
+    def walk_quantifier(self, formula, args):
+        quantifiers, matrix = args[0]
+        qvars = set(v for _, qv in quantifiers for v in qv)
+        nq = set(formula.quantifier_vars()) - qvars
+
+        if len(nq) > 0:
+            if formula.is_exists():
+                return (quantifiers + [(self.mgr.Exists, nq)]), matrix
+            else:
+                return (quantifiers + [(self.mgr.ForAll, nq)]), matrix
+        return quantifiers, matrix
+
+    def walk_theory_op(self, formula, args):
+        return None
+
+    def walk_function(self, formula, args):
+        if formula.function_name().symbol_type().return_type.is_bool_type():
+            return [], formula
+        return None
+
+
+def nnf(formula, environment=None):
+    """Converts the given formula in NNF"""
+    nnfizer = NNFizer(environment)
```

Comment:
Would it make sense to instantiate this lazily in the environment. In this way we could reuse the memoization.